File: February2021/MinimumRemovetoMakeValidParentheses.java
class Solution {

    // TC : O(n)
    // SC :O(n) => "))(("
    public String minRemoveToMakeValid(String s) {
        Stack<Pair<Character, Integer>> st = new Stack<>();
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(c == ')' || c == '('){
                if(st.empty()){
                    st.push(new Pair<>(c, i));
                } else{
                    if(c == ')' && st.peek().getKey() == '('){
                        st.pop();
                    } else{
                        st.push(new Pair<>(c, i));
                    }
                }
            }
        }

        Set<Integer> indexesToBeRemoved = new HashSet<>();
        while(!st.empty()){
            indexesToBeRemoved.add(st.peek().getValue());
            st.pop();
        }

        String ans = "";
        for(int i=0;i<s.length();i++){
            if(!indexesToBeRemoved.contains(i)){
                ans += s.charAt(i);
            }
        }

        return ans;
    }
}

File: March2022/Kotlin/MinimumRemovetoMakeValidParentheses.kt
// Author: Shobhit Behl (LC: shobhitbruh)
class Solution {
     inner class Pair internal constructor(var pos: Int, var x: Char)

    fun minRemoveToMakeValid(s: String): String {
        val st = Stack<Pair>()
        val hs = HashSet<Int>()
        for (i in 0 until s.length) {
            if (s[i] == '(') {
                st.push(Pair(i, s[i]))
            } else if (s[i] == ')') {
                if (st.size > 0 && st.peek().x == '(') {
                    st.pop()
                } else {
                    hs.add(i)
                }
            }
        }
        while (st.size > 0) {
            val m = st.pop()
            hs.add(m.pos)
        }
        val sb = StringBuilder()
        for (i in 0 until s.length) {
            if (!hs.contains(i)) {
                sb.append(s[i])
            }
        }
        return sb.toString()
    }
}